typing/typedecl_unboxed.ml

```diff
@@ -307,16 +306,39 @@ module Head_shape = struct
           let ty = Btype.newgenty (Tconstr (path, params, ref Mnil)) in
           let callstack_map = Callstack.fill ty [] in
           let shape = of_type env ty callstack_map in
-          if !Clflags.dump_headshape then
+          let bound_of_shape = function
+            | Shape_set l -> Types.Max (List.fold_left max 0 l)
+            | Shape_any -> Types.Unbounded
+          in
+          match cstrs with
+            | [] -> invalid_arg "Head_shape.check_typedecl"
+            | cstr :: _ -> begin
+                let vd = cstr.cstr_variants in
+                let imm_bound = bound_of_shape shape.head_imm in
+                let tag_bound = bound_of_shape shape.head_blocks in
+                vd.vd_max_values <- Some (imm_bound, tag_bound)
+              end;
+          if print && !Clflags.dump_headshape then
             Format.fprintf Format.err_formatter "SHAPE(%a) %a@."
-              Ident.print id
+              Path.print path
               pp shape
           end
         with Conflict ->
-          if !Clflags.dump_headshape then
+          if print && !Clflags.dump_headshape then
             Format.fprintf Format.err_formatter "SHAPE(%a) CONFLICT@."
-              Ident.print id
+              Path.print path
       end
     | _ -> ()
 
+  let check_typedecl env (id,decl) = check ~print:true env (id,decl)
+
+  let max_val_of_cstr_descr env cstr =
+    let variant_path = match get_desc cstr.cstr_res with
+      | Tconstr (p, _, _) -> p
+      | _ -> assert false (* XXX *)
+    in
+    let variant_decl = Env.find_type variant_path env in
```

Comment:
I propose to factorize this logic as a helper functions in Types:

```ocaml
val cstr_type_path : constructor_description -> Path.t
```

moving the `assert false` to one central place instead of duplicating it at each use-site of this logic.

Reply:
Note: this could/should be a "refactoring PR" against trunk, also improving many current use-cites for `cstr_res` that are doing exactly this. Would you like to send it, or should I?

Reply:
I'll do it :-)

Reply:
The only problem I see with this refactoring is that the "error handling" is not the same at each use site of `cstr_res`

Reply:
`mark_constructor_description_used` suggests that it is sound to just `assert false`